### PR TITLE
Fix Index Deletion During Partial Snapshot Create

### LINKED ISSUE #50200: Snapshot Finalization Throws NPE in 7.5
We're seeing the following NPE in Cloud logs:

```
Caused by: java.lang.NullPointerException
	at org.elasticsearch.repositories.blobstore.ChecksumBlobStoreFormat.write(ChecksumBlobStoreFormat.java:226) ~[elasticsearch-7.5.0.jar:7.5.0]
	at org.elasticsearch.repositories.blobstore.ChecksumBlobStoreFormat.writeTo(ChecksumBlobStoreFormat.java:197) ~[elasticsearch-7.5.0.jar:7.5.0]
	at org.elasticsearch.repositories.blobstore.ChecksumBlobStoreFormat.write(ChecksumBlobStoreFormat.java:185) ~[elasticsearch-7.5.0.jar:7.5.0]
	at org.elasticsearch.repositories.blobstore.BlobStoreRepository.lambda$finalizeSnapshot$26(BlobStoreRepository.java:738) ~[elasticsearch-7.5.0.jar:7.5.0]
	at org.elasticsearch.action.ActionRunnable$1.doRun(ActionRunnable.java:45) ~[elasticsearch-7.5.0.jar:7.5.0]
	... 5 more
```

during snapshot finalization. This seems like a bug where the cluster metadata does not contain an index for a name that was just snapshot (the deeper cause is that `clusterMetaData.index(index.getName())` in `BlobStoreRepository` returns `null`). I'm investigating this.

### PR DESCRIPTION
**Pull request by original-brownbear**
We can simply filter out shard generation updates for indices
that were removed from the cluster state concurrently to fix
index deletes during partial snapshots as that completely removes
any reference to those shards from the snapshot.

Follow up to #50202
Closes #50200

### CODE REVIEW

**Comment by original-brownbear** [comment]
@ywelsch turns out for `7.6+` this is a really trivial fix. `7.5` requires a different fix though since we use a list of indices to track what indices to finalize instead of the shard generations object.

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/bwc (known Docker issue)

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 1]
```diff
@@ -593,11 +594,22 @@ private void cleanupAfterError(Exception exception) {
...
     }
 
-    private static ShardGenerations buildGenerations(SnapshotsInProgress.Entry snapshot) {
+    private static ShardGenerations buildGenerations(SnapshotsInProgress.Entry snapshot, MetaData metaData) {
         ShardGenerations.Builder builder = ShardGenerations.builder();
         final Map<String, IndexId> indexLookup = new HashMap<>();
-        snapshot.indices().forEach(idx -> indexLookup.put(idx.getName(), idx));
-        snapshot.shards().forEach(c -> builder.put(indexLookup.get(c.key.getIndexName()), c.key.id(), c.value.generation()));
+        snapshot.indices().forEach(idx -> {
+            if (metaData.index(idx.getName()) != null) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: It's a bit annoying that we have to lookup by index name here, not index uuid. In theory, an index could have been deleted and recreated between snapshot start and completion. Do we have index uuid perhaps in the shard snapshots?

**Review by ywelsch — COMMENTED** [review]
How does this affect the "create snapshot" response? Will the index still be returned as part of the response (i.e. returned as successfully snapshotted index)?

**Comment by original-brownbear** [comment]
@ywelsch thanks for tkaing a look.

> How does this affect the "create snapshot" response? Will the index still be returned as part of the response (i.e. returned as successfully snapshotted index)?

The response uses `SnapshotInfo` which is built from `ShardGenerations` so all is well here :) I did in fact find that the `totalShards` was off though and fixed that to be computed from the `ShardGenerations` for partial snapshots as well and it's covered by the new test.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: Good point, this is actually a possibility (I reproduced it with the test added in d219dfa6213) and fixed it now by using the `Index` instances in the shards map from `SnapshotInProgress` :)

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks Yannick!

**Pull request merged**